File: latest/sphinx/snippets/cpp/using/backends/trajectory_batching.cpp
/*******************************************************************************
 * Copyright (c) 2022 - 2026 NVIDIA Corporation & Affiliates.                  *
 * All rights reserved.                                                        *
 *                                                                             *
 * This source code and the accompanying materials are made available under    *
 * the terms of the Apache License 2.0 which accompanies this distribution.    *
 ******************************************************************************/

// [Begin Documentation]
#include <chrono>
#include <cudaq.h>
#include <iostream>

struct xOp {
  void operator()(int qubit_count) __qpu__ {
    cudaq::qvector q(qubit_count);
    x(q);
    mz(q);
  }
};

int main() {
  // Add a simple bit-flip noise channel to X gate
  const double error_probability = 0.01;

  cudaq::bit_flip_channel bit_flip(error_probability);
  // Add noise channels to our noise model.
  cudaq::noise_model noise_model;
  // Apply the bitflip channel to any X-gate on any qubits
  noise_model.add_all_qubit_channel<cudaq::types::x>(bit_flip);

  const int qubit_count = 10;
  const auto start_time = std::chrono::high_resolution_clock::now();
  // Due to the impact of noise, our measurements will no longer be uniformly in
  // the |1...1> state.
  auto counts =
      cudaq::sample({.shots = 1000, .noise = noise_model}, xOp{}, qubit_count);
  const auto end_time = std::chrono::high_resolution_clock::now();
  counts.dump();
  const std::chrono::duration<double, std::milli> elapsed_time =
      end_time - start_time;
  std::cout << "Simulation elapsed time: " << elapsed_time.count() << "ms\n";
  return 0;
}

File: pr-4418/sphinx/snippets/python/using/backends/trajectory_batching.py
# ============================================================================ #
# Copyright (c) 2022 - 2024 NVIDIA Corporation & Affiliates.                   #
# All rights reserved.                                                         #
#                                                                              #
# This source code and the accompanying materials are made available under     #
# the terms of the Apache License 2.0 which accompanies this distribution.     #
# ============================================================================ #

#[Begin Docs]
import time
import cudaq
# Use the `nvidia` target
cudaq.set_target("nvidia")

# Let's define a simple kernel that we will add noise to.
qubit_count = 10


@cudaq.kernel
def kernel(qubit_count: int):
    qvector = cudaq.qvector(qubit_count)
    x(qvector)
    mz(qvector)


# Add a simple bit-flip noise channel to X gate
error_probability = 0.01
bit_flip = cudaq.BitFlipChannel(error_probability)

# Add noise channels to our noise model.
noise_model = cudaq.NoiseModel()
# Apply the bit-flip channel to any X-gate on any qubits
noise_model.add_all_qubit_channel("x", bit_flip)

ideal_counts = cudaq.sample(kernel, qubit_count, shots_count=1000)

start = time.time()
# Due to the impact of noise, our measurements will no longer be uniformly
# in the |1...1> state.
noisy_counts = cudaq.sample(kernel,
                            qubit_count,
                            noise_model=noise_model,
                            shots_count=1000)
end = time.time()
noisy_counts.dump()
print(f"Simulation elapsed time: {(end - start) * 1000} ms")